KAFKA-10585: Kafka Streams should clean up the state store directory from cleanup

dongjinleekr opened the pull request:
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

vvcephei commented:
Thanks for the PR, @dongjinleekr !

Just a couple of meta-questions:
1. Should we have some tests for this change?
2. I think that right now, we'd be leaving behind some task directories and lock files. Maybe, we need to do a little extra legwork to recursively delete the app state directory. An integration test would probably surface this.

dongjinleekr commented:
Hi @vvcephei,

1. Sure. I added some validations in `StateDirectoryTest`.

  - If `StateDirectory#clean` is called, the empty `appDir` is also deleted. (see `StateDirectoryTest#shouldLogManualUserCallMessage`.)
  - If `StateDirectory#clean` is not called, the global state directory and it parent, `appDir` is not deleted. (see `StateDirectoryTest#shouldLogStateDirCleanerMessage`.)

    Please note the difference in `StateDirectoryTest#shouldCleanupAllTaskDirectoriesIncludingGlobalOne`; the `appDir` was an empty directory before but it is now deleted.

2. Since the goal of this modification is deleting the empty directory, we don't need a recursive delete; it is also why I called `File#delete` here, since it works and returns `true` iff the target directory is empty.

And one more thing: I added an exception handling for `SecurityException`. :smile:

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java:
```diff
@@ -510,8 +511,8 @@ public void shouldCleanupAllTaskDirectoriesIncludingGlobalOne() {
-        assertEquals(Collections.emptySet(), Arrays.stream(
-            Objects.requireNonNull(appDir.listFiles())).collect(Collectors.toSet()));
+        // if appDir is empty, it is deleted in StateDirectory#clean process.
+        assertTrue(!appDir.exists());
```
```suggestion
        assertFalse(appDir.exists());
```

:)

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java:
```diff
@@ -602,6 +606,11 @@ public void shouldLogStateDirCleanerMessage() {
             directory.cleanRemovedTasks(cleanupDelayMs);
             assertThat(appender.getMessages(), hasItem(endsWith("ms has elapsed (cleanup delay is " +  cleanupDelayMs + "ms).")));
         }
+
+        // if appDir is empty, it is deleted in  process.
+        // since we did not call StateDirectory#clean, the global state directory is not deleted and appDir also.
+        assertTrue(appDir.exists());
+        assertArrayEquals(appDir.list(), new String[]{"0_0"});
```
We shouldn't add unrelated checks to tests in general, it just makes the tests more confusing. If you want to verify that the directory still exists after closing when we don't call `clean()`, we should just add a new test for it.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java:
```diff
@@ -586,6 +587,9 @@ public void shouldLogManualUserCallMessage() {
                 hasItem(endsWith("as user calling cleanup."))
             );
         }
+
+        // if appDir is empty, it is deleted in StateDirectory#clean process.
+        assertTrue(!appDir.exists());
```
Similar feedback here as below. This test is about logging, not cleanup. You've already added a check to the "shouldCleanup" test, so we don't need one here.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java:
```diff
@@ -301,6 +301,22 @@ public synchronized void clean() {
         }
+
+        try {
+            if (hasPersistentStores && stateDir.exists() && !stateDir.delete()) {
```
```suggestion
            if (stateDir.exists() && !stateDir.delete()) {
```

Do we need to check `hasPersistentStores` here? It seems sufficient just to check if the directory exists.

dongjinleekr replied:
Exactly. But I thought keeping symmetry with the Consturctor is better.

```
if (this.hasPersistentStores && !stateDir.exists() && !stateDir.mkdir()) {
    throw new ProcessorStateException(
        String.format("state directory [%s] doesn't exist and couldn't be created", stateDir.getPath()));
}
```

vvcephei reviewed (commented):
Thanks for the update, @dongjinleekr !

I see now what I missed before in the `cleanRemovedTasksCalledByUser()` method:
```java
// for manual user call, stream threads are not running so it is safe to delete
// the whole directory
Utils.delete(taskDir);
```

So, I agree that for a user-called cleanup, we should indeed see the application directory cleaned up.

If I may offer some suggestions, it seems like we could improve the test coverage by checking that we get the warning and _do not_ delete the state dir if it's not empty. Based on my reading of the code, we should be able to fall through to that point if you set up a test and add an extra file or directory that does not look like a task id. For example just creating a file like "dummy" should prevent the delete and print the warning.

I'm also wondering if we should add an integration test. We have just had so much grief over state directory maintenence, problems with lock files, etc., that I'd feel better with an integration test in place. It should be sufficient to create and start a KafkaStreams application, then block until it goes to running (see `IntegrationTestUtils.getRunningStreams`). Then we'd verify that the application state directory does exist. Then, we'd stop it and block until it's `NOT_RUNNING`. Once it's closed, we should call `clean` and verify that the application state directory is not there anymore.

dongjinleekr commented:
> it seems like we could improve the test coverage by checking that we get the warning and do not delete the state dir if it's not empty.

Totally agree. So, I added two dedicated tests for these cases. See the following test methods:

- `StateDirectoryTest#shouldDeleteAppDirWhenCleanUpIfEmpty`
- `StateDirectoryTest#shouldNotDeleteAppDirWhenCleanUpIfNotEmpty`

I also added an integration test, `StateDirectoryIntegrationTest`. How about this? :smiley:

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/integration/StateDirectoryIntegrationTest.java:
```diff
@@ -0,0 +1,255 @@
+            // Validate application state directory is created.
+            streams.start();
+            try {
+                runningLatch.await(10 * 1000L, TimeUnit.MILLISECONDS);
```
I'd be tempted to give it more time. Experience says that Jenkins will take unreasonably long to perform these operations. How about using the default timeout in IntegrationTestUtils here and below?

vvcephei reviewed (changes requested):
Thanks, @dongjinleekr !

I'd like to increase the timeouts on the integration tests, but otherwise this is ready to merge!

dongjinleekr commented:
@vvcephei Here it is; I updated the timeout to `IntegrationTestUtils.DEFAULT_TIMEOUT` and rebased onto the latest trunk also.

vvcephei reviewed (approved):
Thanks, @dongjinleekr !

A whole bunch of integration tests failed, but it seems unrelated. Possibly environmental, or possibly trunk is broken. I'm re-running the tests, and I'll also test it out locally.

Thanks so much for this contribution!

vvcephei commented:
Hmm, actually I see those same 20 tests fail for me locally. Can you take a look, @dongjinleekr ?

vvcephei commented:
Just for reference, I don't see those failing tests on trunk (67bc4f08feb50ac135a4d8e1d469747102aad3a6).

dongjinleekr commented:
@vvcephei Here it is; I deleted a outdated condition statement in `EOSUncleanShutdownIntegrationTest`.

vvcephei commented:
Thanks for that fix, @dongjinleekr !

Unfortunately, there are still 61 test failures: https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-9414/6/?cloudbees-analytics-link=scm-reporting%2Fstage%2Ffailure#showFailuresLink

Do you mind taking another look?

vvcephei commented:
Hey @dongjinleekr , just a quick check-in: do you plan to take a look at the test failures?

dongjinleekr commented:
Hi @vvcephei,

Sorry for being late. Defending myself, I have been too busy to finalize my project last week, and it ended yesterday. Sure, I am now resolving the broken tests. There are 21 broken tests with the latest trunk, and I just resolved 16 of them. It will be completed soon.

vvcephei commented:
Perfect. Thanks, @dongjinleekr !

No defense is necessary :) I was just wondering if it was still on your radar.

Thanks again for the contribution!

dongjinleekr commented:
Hi @vvcephei,

Here is the fix. The reason for the broken integration tests was: `KafkaStreams#cleanUp` can be called regardless of the `StreamThread`s are terminated.

That is, if `KafkaStreams#cleanUp` is called when the `KafkaStreams` is still running, the `{state-store-directory}/{application-id}` directory is deleted and  `StreamThread` crashes with the exception from `StateDirectory#directoryForTask` - since it fails to create the task state store directory, i.e., `{state-store-directory}/{application-id}/{task-id}`.

Moreover, `StateDirectory#directoryForTask` method has two additional vulnerabilities:

1. When it creates the task state store directory, it does not create its parent directories automatically - if there is not `{state-store-directory}/{application-id}`, `taskDir.mkdir()` returns false and it throws an exception. For this behavior breaks the integration tests, I modified `taskDir.mkdir()` to `taskDir.mkdirs()` to create `{state-store-directory}/{application-id}` automatically. 
2. This method only checks whether there is a `File` at `{state-store-directory}/{application-id}/{task-id}`, regardless of it is actually a directory or not. I added an additional check condition for this case and `StateDirectoryTest#shouldThrowProcessorStateException` is updated accordingly.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/integration/EOSUncleanShutdownIntegrationTest.java:
```diff
@@ -140,9 +140,6 @@ public void shouldWorkWithUncleanShutdownWipeOutStateStore() throws InterruptedE
             IntegrationTestUtils.produceSynchronously(producerConfig, false, input, Optional.empty(),
                 singletonList(new KeyValueTimestamp<>("k1", "v1", 0L)));
 
-            TestUtils.waitForCondition(stateDir::exists,
-                "Failed awaiting CreateTopics first request failure");
```
Can you explain why we need to remove this? It seems like the application must have created the state directory by this point, right?

dongjinleekr replied:
- Previous: The test asserts that the (empty) StateStore directory is not deleted.
- Now: The empty StateStore directory is deleted in the cleanup process, so this assertion is no longer valid. (wait, would it much better to negate the logical condition instead of removing it?)

vvcephei replied:
Hmm, I see. Maybe I misread it. I thought we were asserting that Streams actually creates the directory when it starts processing, so that the later assertion (L159) that it cleans up the directory is actually valid.

Then again, L159 seems to assert the opposite thing that its own comment states, so maybe this whole test is just wacky.

The only clue about what we're trying to do here is the name, which almost doesn't make sense at all: `shouldWorkWithUncleanShutdownWipeOutStateStore`. I guess it's saying that we should delete the state store on unclean shutdown? I think that's not what we do anyway. I think we just guarantee that we write an empty checkpoint file. So maybe we should instead change the assertion to:
* the directory is absent
* OR the directory is present, but the checkpoint file is missing
* OR the directory and checkpoint file are present, but the checkpoint file is empty

Any of those should indicate a successful unclean shutdown.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java:
```diff
@@ -186,19 +186,29 @@ public void shouldReportDirectoryEmpty() throws IOException {
 
     @Test
     public void shouldThrowProcessorStateException() throws IOException {
```
Since you have modified the purpose of this test, maybe we can go ahead and give the test a more specific name as well. 

```suggestion
    public void shouldThrowProcessorStateExceptionIfTaskDirectoryIsOccupiedByFile() throws IOException {
```

Also, I won't dispute the value of checking this condition, but would like to point out that this test was previously verifying a specific error on failure to create the task directory, and now we are no longer checking that failure. In other words, we were previously verifying "task directory [%s] doesn't exist and couldn't be created", but now we are only verifying the separate and specific failure reason "task directory path [%s] is already occupied".

It actually seems like maybe we don't need to check that specific `!taskDir.isDirectory()` case, since it seems like having this file sitting there should cause a failure to create the task directory, right?

dongjinleekr replied:
Totally agree. :smile:

vvcephei reviewed (commented):
Hey, thanks for that update, @dongjinleekr !

Sorry to kick this back to you one last time, but I had some questions about that new branch and some of the test edits. State directory management is a particularly tricky bunch of logic, so I just want to make sure we're not weakening our test coverage here.

dongjinleekr commented:
Hi @vvcephei,

I re-based the branch onto the latest trunk, reorganized the changes into two commits, and fixed `EOSUncleanShutdownIntegrationTest#shouldWorkWithUncleanShutdownWipeOutStateStore` - Sure, its contents do not follow the description of its name. :wink:

dongjinleekr commented:
Retest this please.

dongjinleekr commented:
Rebased onto the latest trunk.

dongjinleekr commented:
Rebased onto the latest trunk. cc/ @vvcephei @mjsax

dongjinleekr committed:
Make application's statestore directory is deleted in cleanup process if it is empty

1. Update StateDirectory#clean

  - Delete application's statestore directory in cleanup process if it is empty.

2. Add Tests

  - StateDirectoryTest#shouldDeleteAppDirWhenCleanUpIfEmpty: asserting the empty application directory is deleted with StateDirectory#clean.
  - StateDirectoryTest#shouldNotDeleteAppDirWhenCleanUpIfNotEmpty: asserting the non-empty application directory is not deleted with StateDirectory#clean and appropriate log message is generated.
  - Add Integration test: StateDirectoryIntegrationTest

3. Improve EOSUncleanShutdownIntegrationTest: test all available cases regarding cleanup process on unclean shutdown.

dongjinleekr commented:
Hi @vvcephei,

It seems like I have failed to understand your intention correctly. Here is the fix. I reorganized the PR into two commits, one for `StateDirectory` and the other one for `EOSUncleanShutdownIntegrationTest`. The following summarizes the updates:

1. Rebase the PR onto the latest trunk, resolving all broken syntaxes.
2. Remove redundant assertion in `StateDirectoryTest#shouldNotDeleteAppDirWhenCleanUpIfNotEmpty`; it now validates logging message only.
3. Make `StateDirectoryIntegrationTest` to use default timeout (`IntegrationTestUtils.DEFAULT_TIMEOUT`) with its latches, keeping the suite from being flaky.

dongjinleekr commented:
Add to this, I also updated `EOSUncleanShutdownIntegrationTest` also. I inspected the history of this test suite precisely and found the following:

1. This suite was introduced in c2ec974 by @abbccdda. It asserts that the Task's statestore directory (not root one in KafkaStreams's `state.dir` property) is deleted after an unclean shutdown. However, IMHO this assertion is a little bit inaccurate since the Task's statestore directory may not be deleted, for example, with an empty checkpoint file.
2. A comment is added on L160 in d3c067f by @guozhangwang - 'the state directory should still exist with the empty checkpoint file'. However, IMHO this comment seems to be a mistake, since `assertFalse(stateDir.exists());` on L161 asserts that the Task's statestore directory does not exist.

There are some other commits on this file later, but all of them are just minor improvements, not changing the testing logic.

With this PR, Kafka Streams now deletes the empty application statestore directory (i.e., `{state.dir}/{application.id}`). So, I improved this suite like this:

1. After initialization, wait until the Task's statestore directory is populated. (i.e., `{state.dir}/{application.id}/0_0/*`)
2. Assert that one of the following is satisfied:

  - The Task's statestore directory is empty, so deleted.
  - The Task's statestore directory is not empty but without a checkpoint file.
  - The Task's statestore directory is not empty but with an empty checkpoint file.

(Wait, then should we separate the modifications on `EOSUncleanShutdownIntegrationTest` into an independent PR?)

Please have a look when you are free. Thanks again for reviewing my PR! :smiley:

guozhangwang commented:
@dongjinleekr @vvcephei just checking are you still working on it? I saw @vvcephei has approved the PR but it was not yet merged, hence asking.

dongjinleekr commented:
Hi @guozhangwang,

Thanks for your mention. I completed applying @vvcephei's comment (on March 25, 2021) and awaiting review and merging, with maintaining the PR to the latest trunk.

As you can see here, this issue was started much earlier but It seems like I misunderstood @vvcephei's intention. 😢 And I recognized it this March.

guozhangwang commented:
@dongjinleekr just FYI, I know @vvcephei was quite swamped in the past months with pretty heavy release management duties -- you'd see his email about 2.8.0 release, which was all thanks to him -- so I believe it's just he was too busy with 2.8.0 to review your PRs, and I'd like to thank you for your patience!

@vvcephei whenever you have time please let know if the current PR is good or not.

dongjinleekr commented:
> I know @vvcephei was quite swamped in the past months with pretty heavy release management duties.

Totally agree. It's also why I did not press him. :smile: I rebased the PR into the latest trunk and checked it passes all tests!

dongjinleekr commented:
Hi @vvcephei, Could you review this PR now? :bow:

guozhangwang reviewed (approved):
I made a quick pass on the non-testing code and it looks good to me. cc @vvcephei for a final look and merge.

dongjinleekr commented:
Hi @guozhangwang @vvcephei,
Could you have a look now? :smiley:

dongjinleekr commented:
Rebased onto the latest trunk. cc/ @vvcephei

vvcephei reviewed (approved):
Thanks for the update @dongjinleekr ; sorry it took so long to get this merged.

vvcephei commented:
It looks like the core integration tests have gotten into bad shape. They've been failing on trunk as well. I just ran the Streams integration tests on my machine, and they passed, so I'll go ahead and merge.
